src/formats/jpeg.zig

```diff
@@ -0,0 +1,1084 @@
+    // Start of Frame markers, non-differential, arithmetic coding
+    SOF9 = 0xFFC9, // Extended sequential DCT
+    SOF10 = 0xFFCA, // Progressive DCT
+    SOF11 = 0xFFCB, // Lossless sequential
+    // Start of Frame markers, differential, arithmetic coding
+    SOF13 = 0xFFCD, // Differential sequential DCT
+    SOF14 = 0xFFCE, // Differential progressive DCT
+    SOF15 = 0xFFCF, // Differential lossless sequential
+
+    DefineHuffmanTables = 0xFFC4,
+    DefineArithmeticCoding = 0xFFCC,
+
+    StartOfImage = 0xFFD8,
+    EndOfImage = 0xFFD9,
+    StartOfScan = 0xFFDA,
+    DefineQuantizationTables = 0xFFDB,
+    DefineNumberOfLines = 0xFFDC,
+    DefineRestartInterval = 0xFFDD,
+    DefineHierarchicalProgression = 0xFFDE,
+    ExpandReferenceComponents = 0xFFDF,
+
+    // Add 0-15 as needed.
+    Application0 = 0xFFE0,
+    // Add 0-13 as needed.
+    JpegExtension0 = 0xFFF0,
+    Comment = 0xFFFE,
+};
+
+const DensityUnit = enum {
+    Pixels,
+    DotsPerInch,
+    DotsPerCm,
+};
+
+const JFIFHeader = struct {
+    jfifRevision: u16,
+    densityUnit: DensityUnit,
+    xDensity: u16,
+    yDensity: u16,
+
+    fn init(reader: ImageReader, seek_stream: ImageSeekStream) !JFIFHeader {
+        // Read the first APP0 header.
+        try seek_stream.seekTo(2);
+        const maybe_app0_marker = try reader.readIntBig(u16);
+        if (maybe_app0_marker != @enumToInt(Markers.Application0)) {
+            return error.App0MarkerDoesNotExist;
+        }
+
+        // Header length
+        _ = try reader.readIntBig(u16);
+
+        var identifier_buffer: [4]u8 = undefined;
+        _ = try reader.read(identifier_buffer[0..]);
+
+        if (!std.mem.eql(u8, identifier_buffer[0..], "JFIF")) {
+            return error.JfifIdentifierNotSet;
+        }
+
+        // NUL byte after JFIF
+        _ = try reader.readByte();
+
+        const jfifRevision = try reader.readIntBig(u16);
+        const densityUnit = @intToEnum(DensityUnit, try reader.readByte());
+        const xDensity = try reader.readIntBig(u16);
+        const yDensity = try reader.readIntBig(u16);
+
+        const thumbnailWidth = try reader.readByte();
+        const thumbnailHeight = try reader.readByte();
+
+        if (thumbnailWidth != 0 or thumbnailHeight != 0) {
+            return error.ThumbnailImagesUnsupported;
+        }
+
+        // Make sure there are no application markers after us.
+        if (((try reader.readIntBig(u16)) & 0xFFF0) == @enumToInt(Markers.Application0)) {
+            return error.ExtraneousApplicationMarker;
+        }
+
+        try seek_stream.seekBy(-2);
+
+        return JFIFHeader{
+            .jfifRevision = jfifRevision,
+            .densityUnit = densityUnit,
+            .xDensity = xDensity,
+            .yDensity = yDensity,
+        };
+    }
+};
+
+const QuantizationTable = union(enum) {
+    Q8: [64]u8,
+    Q16: [64]u16,
+
+    pub fn init(precision: u8, reader: ImageReader) !QuantizationTable {
```

Comment:
Same as below, should be `read` instead of `init`

Reply:
Done.